File: SignalServiceKit/src/TestUtils/MockSSKEnvironment.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "MockSSKEnvironment.h"
#import "OWS2FAManager.h"
#import "OWSBackgroundTask.h"
#import "OWSDisappearingMessagesJob.h"
#import "OWSFakeProfileManager.h"
#import "OWSIdentityManager.h"
#import "OWSMessageManager.h"
#import "OWSOutgoingReceiptManager.h"
#import "OWSReceiptManager.h"
#import "ProfileManagerProtocol.h"
#import "SSKPreKeyStore.h"
#import "SSKSignedPreKeyStore.h"
#import "StorageCoordinator.h"
#import "TSAccountManager.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

#ifdef TESTABLE_BUILD

@implementation MockSSKEnvironment

- (instancetype)init
{
    // Ensure that OWSBackgroundTaskManager is created now.
    [OWSBackgroundTaskManager shared];

    StorageCoordinator *storageCoordinator = [StorageCoordinator new];
    SDSDatabaseStorage *databaseStorage = storageCoordinator.databaseStorage;

    // Set up DependenciesBridge

    AccountServiceClient *accountServiceClient = [FakeAccountServiceClient new];
    OWSIdentityManager *identityManager = [[OWSIdentityManager alloc] initWithDatabaseStorage:databaseStorage];
    MessageProcessor *messageProcessor = [MessageProcessor new];
    MessageSender *messageSender = [FakeMessageSender new];
    NetworkManager *networkManager = [OWSFakeNetworkManager new];
    OWS2FAManager *ows2FAManager = [OWS2FAManager new];
    SignalProtocolStore *pniSignalProtocolStore = [[SignalProtocolStore alloc] initForIdentity:OWSIdentityPNI];
    id<OWSSignalServiceProtocol> signalService = [OWSSignalServiceMock new];
    OWSFakeStorageServiceManager *storageServiceManager = [OWSFakeStorageServiceManager new];
    id<SyncManagerProtocol> syncManager = [[OWSMockSyncManager alloc] init];
    TSAccountManager *tsAccountManager = [TSAccountManager new];

    [DependenciesBridgeSetup setupSingletonWithAccountServiceClient:accountServiceClient
                                                    databaseStorage:databaseStorage
                                                    identityManager:identityManager
                                                   messageProcessor:messageProcessor
                                                      messageSender:messageSender
                                                     networkManager:networkManager
                                                      ows2FAManager:ows2FAManager
                                                   pniProtocolStore:pniSignalProtocolStore
                                                      signalService:signalService
                                              storageServiceManager:storageServiceManager
                                                        syncManager:syncManager
                                                   tsAccountManager:tsAccountManager];

    // Set up ourselves

    id<ContactsManagerProtocol> contactsManager = [OWSFakeContactsManager new];
    OWSLinkPreviewManager *linkPreviewManager = [OWSLinkPreviewManager new];
    id<PendingReceiptRecorder> pendingReceiptRecorder = [NoopPendingReceiptRecorder new];
    id<ProfileManagerProtocol> profileManager = [OWSFakeProfileManager new];
    OWSMessageManager *messageManager = [OWSMessageManager new];
    BlockingManager *blockingManager = [BlockingManager new];
    id<RemoteConfigManager> remoteConfigManager = [StubbableRemoteConfigManager new];
    SignalProtocolStore *aciSignalProtocolStore = [[SignalProtocolStore alloc] initForIdentity:OWSIdentityACI];
    id<OWSUDManager> udManager = [OWSUDManagerImpl new];
    OWSMessageDecrypter *messageDecrypter = [OWSMessageDecrypter new];
    GroupsV2MessageProcessor *groupsV2MessageProcessor = [GroupsV2MessageProcessor new];
    SocketManager *socketManager = [[SocketManager alloc] init];
    OWSDisappearingMessagesJob *disappearingMessagesJob = [OWSDisappearingMessagesJob new];
    OWSReceiptManager *receiptManager = [OWSReceiptManager new];
    OWSOutgoingReceiptManager *outgoingReceiptManager = [OWSOutgoingReceiptManager new];
    id<SSKReachabilityManager> reachabilityManager = [MockSSKReachabilityManager new];
    id<OWSTypingIndicators> typingIndicators = [[OWSTypingIndicatorsImpl alloc] init];
    OWSAttachmentDownloads *attachmentDownloads = [[OWSAttachmentDownloads alloc] init];
    StickerManager *stickerManager = [[StickerManager alloc] init];
    SignalServiceAddressCache *signalServiceAddressCache = [SignalServiceAddressCache new];
    SSKPreferences *sskPreferences = [SSKPreferences new];
    id<GroupsV2> groupsV2 = [[MockGroupsV2 alloc] init];
    id<GroupV2Updates> groupV2Updates = [[MockGroupV2Updates alloc] init];
    MessageFetcherJob *messageFetcherJob = [MessageFetcherJob new];
    BulkProfileFetch *bulkProfileFetch = [BulkProfileFetch new];
    id<VersionedProfiles> versionedProfiles = [MockVersionedProfiles new];
    ModelReadCaches *modelReadCaches =
        [[ModelReadCaches alloc] initWithModelReadCacheFactory:[TestableModelReadCacheFactory new]];
    EarlyMessageManager *earlyMessageManager = [EarlyMessageManager new];
    OWSMessagePipelineSupervisor *messagePipelineSupervisor = [OWSMessagePipelineSupervisor createStandardSupervisor];
    AppExpiry *appExpiry = [AppExpiry new];
    id<PaymentsHelper> paymentsHelper = [MockPaymentsHelper new];
    id<PaymentsCurrencies> paymentsCurrencies = [MockPaymentsCurrencies new];
    id<PaymentsEvents> paymentsEvents = [PaymentsEventsNoop new];
    id<MobileCoinHelper> mobileCoinHelper = [MobileCoinHelperMock new];
    SpamChallengeResolver *spamChallengeResolver = [SpamChallengeResolver new];
    SenderKeyStore *senderKeyStore = [SenderKeyStore new];
    PhoneNumberUtil *phoneNumberUtil = [PhoneNumberUtil new];
    id webSocketFactory = [WebSocketFactoryMock new];
    LegacyChangePhoneNumber *legacyChangePhoneNumber = [LegacyChangePhoneNumber new];
    id<SubscriptionManager> subscriptionManager = [MockSubscriptionManager new];
    SystemStoryManagerMock *systemStoryManager = [SystemStoryManagerMock new];
    RemoteMegaphoneFetcher *remoteMegaphoneFetcher = [RemoteMegaphoneFetcher new];
    SSKJobQueues *sskJobQueues = [SSKJobQueues new];
    id contactDiscoveryManager = [ContactDiscoveryManagerImpl new];

    self = [super initWithContactsManager:contactsManager
                       linkPreviewManager:linkPreviewManager
                            messageSender:messageSender
                   pendingReceiptRecorder:pendingReceiptRecorder
                           profileManager:profileManager
                           networkManager:networkManager
                           messageManager:messageManager
                          blockingManager:blockingManager
                          identityManager:identityManager
                      remoteConfigManager:remoteConfigManager
                   aciSignalProtocolStore:aciSignalProtocolStore
                   pniSignalProtocolStore:pniSignalProtocolStore
                                udManager:udManager
                         messageDecrypter:messageDecrypter
                 groupsV2MessageProcessor:groupsV2MessageProcessor
                            socketManager:socketManager
                         tsAccountManager:tsAccountManager
                            ows2FAManager:ows2FAManager
                  disappearingMessagesJob:disappearingMessagesJob
                           receiptManager:receiptManager
                   outgoingReceiptManager:outgoingReceiptManager
                      reachabilityManager:reachabilityManager
                              syncManager:syncManager
                         typingIndicators:typingIndicators
                      attachmentDownloads:attachmentDownloads
                           stickerManager:stickerManager
                          databaseStorage:databaseStorage
                signalServiceAddressCache:signalServiceAddressCache
                            signalService:signalService
                     accountServiceClient:accountServiceClient
                    storageServiceManager:storageServiceManager
                       storageCoordinator:storageCoordinator
                           sskPreferences:sskPreferences
                                 groupsV2:groupsV2
                           groupV2Updates:groupV2Updates
                        messageFetcherJob:messageFetcherJob
                         bulkProfileFetch:bulkProfileFetch
                        versionedProfiles:versionedProfiles
                          modelReadCaches:modelReadCaches
                      earlyMessageManager:earlyMessageManager
                messagePipelineSupervisor:messagePipelineSupervisor
                                appExpiry:appExpiry
                         messageProcessor:messageProcessor
                           paymentsHelper:paymentsHelper
                       paymentsCurrencies:paymentsCurrencies
                           paymentsEvents:paymentsEvents
                         mobileCoinHelper:mobileCoinHelper
                    spamChallengeResolver:spamChallengeResolver
                           senderKeyStore:senderKeyStore
                          phoneNumberUtil:phoneNumberUtil
                         webSocketFactory:webSocketFactory
                  legacyChangePhoneNumber:legacyChangePhoneNumber
                      subscriptionManager:subscriptionManager
                       systemStoryManager:systemStoryManager
                   remoteMegaphoneFetcher:remoteMegaphoneFetcher
                             sskJobQueues:sskJobQueues
                  contactDiscoveryManager:contactDiscoveryManager];

    if (!self) {
        return nil;
    }

    self.callMessageHandlerRef = [OWSFakeCallMessageHandler new];
    self.notificationsManagerRef = [NoopNotificationsManager new];

    return self;
}

- (void)setContactsManagerForMockEnvironment:(id<ContactsManagerProtocol>)contactsManager
{
    [super setContactsManagerRef:contactsManager];
}

@end

#endif

NS_ASSUME_NONNULL_END

File: SignalServiceKit/TestUtils/MockSSKEnvironment.swift
//
// Copyright 2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import GRDB

#if TESTABLE_BUILD

public class MockSSKEnvironment {
    /// Set up a mock SSK environment as well as ``DependenciesBridge``.
    @MainActor
    public static func activate(
        appReadiness: any AppReadiness = AppReadinessImpl(),
        callMessageHandler: any CallMessageHandler = NoopCallMessageHandler(),
        currentCallProvider: any CurrentCallProvider = CurrentCallNoOpProvider(),
        notificationPresenter: any NotificationPresenter = NoopNotificationPresenterImpl(),
        testDependencies: AppSetup.TestDependencies? = nil,
    ) async {
        let sampleDatabase = await initializeSampleDatabase()
        _ = await _activate(
            appReadiness: appReadiness,
            callMessageHandler: callMessageHandler,
            currentCallProvider: currentCallProvider,
            notificationPresenter: notificationPresenter,
            testDependencies: testDependencies,
            sampleDatabase: sampleDatabase,
        )
    }

    @MainActor
    private static func _activate(
        appReadiness: any AppReadiness = AppReadinessImpl(),
        callMessageHandler: any CallMessageHandler = NoopCallMessageHandler(),
        currentCallProvider: any CurrentCallProvider = CurrentCallNoOpProvider(),
        keychainStorage: MockKeychainStorage = MockKeychainStorage(),
        notificationPresenter: any NotificationPresenter = NoopNotificationPresenterImpl(),
        testDependencies: AppSetup.TestDependencies? = nil,
        sampleDatabase: SampleDatabase?,
    ) async -> SampleDatabase {
        owsPrecondition(!(CurrentAppContext() is TestAppContext))
        owsPrecondition(!SSKEnvironment.hasShared)
        owsPrecondition(!DependenciesBridge.hasShared)

        let testAppContext = TestAppContext()
        SetCurrentAppContext(testAppContext, isRunningTests: true)

        /// Note that ``SDSDatabaseStorage/grdbDatabaseFileUrl``, through a few
        /// layers of abstraction, uses the "current app context" to decide
        /// where to put the database,
        ///
        /// For a ``TestAppContext`` as configured above, this will be a
        /// subdirectory of our temp directory unique to the instantiation of
        /// the app context.
        let databaseUrl = SDSDatabaseStorage.grdbDatabaseFileUrl

        let keychainStorage: MockKeychainStorage
        if let sampleDatabase {
            sampleDatabase.copyTo(databaseUrl)
            keychainStorage = sampleDatabase.keychainStorage.clone()
        } else {
            keychainStorage = MockKeychainStorage()
        }

        let finalContinuation = await AppSetup().start(
            appContext: testAppContext,
            databaseStorage: try! SDSDatabaseStorage(
                appReadiness: appReadiness,
                databaseFileUrl: databaseUrl,
                keychainStorage: keychainStorage,
            ),
        ).migrateDatabaseSchema().initGlobals(
            appReadiness: appReadiness,
            backupArchiveErrorPresenterFactory: NoOpBackupArchiveErrorPresenterFactory(),
            deviceBatteryLevelManager: nil,
            deviceSleepManager: nil,
            paymentsEvents: PaymentsEventsNoop(),
            mobileCoinHelper: MobileCoinHelperMock(),
            callMessageHandler: callMessageHandler,
            currentCallProvider: currentCallProvider,
            notificationPresenter: notificationPresenter,
            testDependencies: testDependencies ?? AppSetup.TestDependencies(
                contactManager: FakeContactsManager(),
                groupV2Updates: MockGroupV2Updates(),
                groupsV2: MockGroupsV2(),
                messageSender: { FakeMessageSender(accountChecker: $0) },
                networkManager: OWSFakeNetworkManager(appReadiness: appReadiness, libsignalNet: nil),
                paymentsCurrencies: MockPaymentsCurrencies(),
                paymentsHelper: MockPaymentsHelper(),
                pendingReceiptRecorder: NoopPendingReceiptRecorder(),
                profileManager: OWSFakeProfileManager(),
                reachabilityManager: MockSSKReachabilityManager(),
                remoteConfigManager: StubbableRemoteConfigManager(),
                signalService: OWSSignalServiceMock(),
                storageServiceManager: FakeStorageServiceManager(),
                syncManager: OWSMockSyncManager(),
                systemStoryManager: SystemStoryManagerMock(),
                versionedProfiles: MockVersionedProfiles(),
                webSocketFactory: WebSocketFactoryMock(),
            ),
        ).migrateDatabaseData()
        finalContinuation.runLaunchTasksIfNeededAndReloadCaches()
        return SampleDatabase(fileUrl: databaseUrl, keychainStorage: keychainStorage)
    }

    struct SampleDatabase {
        var fileUrl: URL
        var keychainStorage: MockKeychainStorage

        func copyTo(_ databaseUrl: URL) {
            try! FileManager.default.copyItem(at: self.fileUrl, to: databaseUrl)
        }
    }

    @MainActor
    private static var sampleDatabase: SampleDatabase?

    @MainActor
    private static func initializeSampleDatabase() async -> SampleDatabase {
        if let sampleDatabase {
            return sampleDatabase
        }
        let oldContext = CurrentAppContext()
        let result = await MockSSKEnvironment._activate(sampleDatabase: nil)
        try! SSKEnvironment.shared.databaseStorageRef.grdbStorage.syncTruncatingCheckpoint()
        self.sampleDatabase = result
        await MockSSKEnvironment.deactivateAsync(oldContext: oldContext)
        return result
    }

    @MainActor
    private static func flushAndWait() {
        AssertIsOnMainThread()

        waitForMainQueue()

        // Wait for all pending readers/writers to finish.
        SSKEnvironment.shared.databaseStorageRef.grdbStorage.pool.barrierWriteWithoutTransaction { _ in }

        // Wait for the MessageProcessor to finish.
        SSKEnvironment.shared.messageProcessorRef.serialQueueForTests.sync {}

        // Wait for the main queue *again* in case more work was scheduled.
        waitForMainQueue()
    }

    public static func deactivateAsync(oldContext: any AppContext) async {
        await withCheckedContinuation { continuation in
            DispatchQueue.main.async {
                SSKEnvironment.shared.databaseStorageRef.grdbStorage.pool.barrierWriteWithoutTransaction { _ in }
                SSKEnvironment.shared.messageProcessorRef.serialQueueForTests.async {
                    DispatchQueue.main.async {
                        continuation.resume()
                    }
                }
            }
        }
        _deactivate(oldContext: oldContext)
    }

    @MainActor
    public static func deactivate(oldContext: any AppContext) {
        flushAndWait()
        _deactivate(oldContext: oldContext)
    }

    private static func _deactivate(oldContext: any AppContext) {
        SetCurrentAppContext(oldContext, isRunningTests: true)
        SSKEnvironment.setShared(nil, isRunningTests: true)
        DependenciesBridge.setShared(nil, isRunningTests: true)
    }

    private static func waitForMainQueue() {
        // Spin the main run loop to flush any remaining async work.
        var done = false
        DispatchQueue.main.async { done = true }
        while !done {
            CFRunLoopRunInMode(.defaultMode, 0.0, true)
        }
    }
}

#endif
